etcdserver/metrics.go

```diff
@@ -32,10 +38,36 @@ var (
 		Name: "etcdserver_proposal_failed_total",
 		Help: "The total number of failed proposals.",
 	})
+
+	fileDescriptorInUse = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "file_descriptor_in_use_total",
+		Help: "The total number of file descriptors in use.",
+	})
 )
 
 func init() {
 	prometheus.MustRegister(proposeDurations)
 	prometheus.MustRegister(proposePending)
 	prometheus.MustRegister(proposeFailed)
+	prometheus.MustRegister(fileDescriptorInUse)
+}
+
+func monitorFileDescriptor(done <-chan struct{}) {
+	limit, _ := runtime.FDLimit()
+	ticker := time.NewTicker(time.Second)
+	defer ticker.Stop()
+	for {
+		cur, err := runtime.FDInUse()
+		if err == nil {
+			fileDescriptorInUse.Set(float64(cur))
+		}
+		if limit != 0 && cur >= limit/5*4 {
+			log.Printf("etcdserver: more than 80 percent file descriptors are in use [inuse = %d, limit = %d]", cur, limit)
```

Comment:
"etcdserver: 80% of the file descriptor limit is open"

Also, it would be nice to have the limit be a const and to be used in printing the string too.